### DISCUSSION TITLE
Sequential Initial Sync

### DISCUSSION THREAD

**Discussion by ollitheorca** [microsoft/Dual-write-automations#6] [Q&A] [answered]
Hi,

We have a Dual write solution that requires our tables to be run/initiated in a particular order to have the correct related data available for the next table map sync.

Is this done simply by editing the config file to run in "sequential" instead of "parallel"? Or, should I also utilize the "groups" to set up what group of tables that can be run first, second, third, etc.?
Also, I can't seem to find in the documentation, if the keyword "sequential" is the correct word to use in the config file under "executionMode"? 

Cheers!

**Comment by Hax1337** [comment]
@ollitheorca Inital sync will always run sequential due to the dependencies. 

If you keep it as "Parallel" the steps executed before the inital sync will be run in parallel, otherwise it will be sequential thats the only difference. 

For the sequence you don't neccesarily need to have different groups, you can use one for all the maps.
The groups control if the inital sync will be executed or not, so you need to have at least 1 group assinged to the maps you want to sync.

The sequence whats taken is the one you see in your configuration, the first map will be synced first. 

Does this answer the question?

**Comment by ollitheorca** [comment]
@Hax1337, thanks a lot for your reply.

It does make sense, that the initial sync part always runs "sequential".

Then I just have one additional question about that. 
When using the standard DW UI and you select some random DW maps in the list, DW will automatically detect the appropriate order of execution due to the relationships of the tables for the DW maps select. Does this tool do the same, or will I need to edit my config file to order my maps in the correct order, to work properly?

Because, I think I can see in my log file, that some DW maps are "complaining" about not have proper lookup values available - that to me could indicate an incorrect execution order, yes? :-)

Cheers!

**Comment by Hax1337** [comment, marked as answer]
@ollitheorca 

At this stage it would not detect the order for it. During Export it does that and brings it into the correct order.

But during execution it follows top to bottom, means if you change the order in the config file you would need to make sure that its the correct order.

What you can do is, run only the maps you want to have in your config and then run "Export config" with "Running" status only.
This will bring you then only the maps you want into your config in the correct order.

**Reply by ollitheorca** [reply]
Thanks a lot for clarifying.

One additional question about the config file - after exporting my config file, all the mappings didn't have any keys set up in the config file. They are all just like this:

`<Map master="FO" mapName="flx_leasetypetables - Billing types – export to Customer Engagement (CE)"
        version="2023.1.0.1" group="All" authors="FlexProperty" keys="" />`

We have entity keys set up on the tables in CE, so I guess they will be set automatically when the tool does a "refresh table" before initializing the sync, yes? Or should I manually set them in the config file? 

Cheers!

**Reply by ollitheorca** [reply]
@Hax1337 Hi again,

This is JFY. :-)

I double-checked the ordering and it seems the config file is not placing the maps in the correct order after all.

Here is an example of two tables that are related - when initializing the "child" table first in the DW UI it shows like this(which is correct):
![image](https://github.com/microsoft/Dual-write-automations/assets/129459105/3c3e7805-7cde-4451-8ee8-6270b42e1d0b)

However, when looking at the config file exported, the maps are put like this:
![image](https://github.com/microsoft/Dual-write-automations/assets/129459105/d4f1e9f1-61d6-4748-99e6-dcbcf228e57f)

Please let me know if I can provide any additional information. 👍

**Reply by Hax1337** [reply]
@ollitheorca  thanks for the information. 

Can it be that they have a relation to each other? Similar to e.g. Customers and the Invoiceaccount, it's like a cross reference scenario where the same table refereces to the same table? 

If its not, I'd like to find out more.
